Adding an integration for Google Chrome Lighthouse web performance audits

ericmustin opened the pull request:
@ruthnaebeck @ofek 
Apologies on the email clutter, this PR implements feedback from https://github.com/DataDog/integrations-extras/pull/351

### What does this PR do?

Adds a check that generates performance, seo, accessibility, progressive web application, and best-practice score metrics from Google Chrome Lighthouse web performance audits. This will basically run the [lighthouse node package](https://developers.google.com/web/tools/lighthouse/#cli) for the urls you specify with chrome in headless mode, then output the ligthouse report in json and collect some baseline metrics from that report.

### Motivation

> What inspired you to submit this pull request?

There's no web performance, seo, accessibility, etc etc information in the current http_check in integrations-core and I wanted to monitor these metrics. Seems like a common use case. Lighthouse seems to be the standard for this within the web performance community

### Review checklist

- [X] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [X] Feature or bugfix has tests
- [ ] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

> Anything else we should know when reviewing?

Hey all, thanks for taking a look at my work.  I've spent a bit of time on this but it definitely could be improved upon, but I wanted to get a PR in early and get some feedback on whether y'all would be open to considering this as a community integration. 

There's a few things here I wasn't sure about conventions on, specifically, 

- The system requirements include installing Node and some npm packages, so I wasn't sure how to convey that but I tried to document that in the readme.

- The reports generated here by lighthouse are very configurable and have some very robust audit details, but for now just including the 5 top level category scores as metrics.

- Also, it's a somewhat resource intensive and long running (anecdotally ~5-10s) sub process being called in the check since it's spinning up a headless chrome browser, I've only tested this with a few example instances and a 60 second collection interval, which seems to work fine and doesn't cause any issues, but I wasn't sure whether it would be best to hard code some limitations in here or the recommended way to approach that.
 
- Lastly, along those lines I am definitely pretty new to python so I tried to conform to best practices here and add some decently robust tests and fixtures, but please feel free to let me know if there's any code quality concerns or feedback here. Happy to put in the work on this and any feedback is much appreciated.

Thanks again for taking a look

ruthnaebeck commented on lighthouse/README.md:
```diff
@@ -0,0 +1,81 @@
+1. Create a `lighthouse.d/` folder in the `conf.d/` folder at the root of your Agent's directory. 
+2. Create a `conf.yaml` file in the `lighthouse.d/` folder previously created.
+3. Reference the example lighthouse [`conf.yaml.example` file][4] in this repository and copy it's contents into the `conf.yaml` file just created.
+4. Edit the `conf.yaml` file for your use case.  **Note**: It's unclear how resource intensive this can be, a typical lighthouse report takes 5-10seconds to generate so it may be necessary to have a higher than usual `minimum_collection_interval`.
```
```suggestion
4. Edit the `conf.yaml` file for your use case.  **Note**: It's unclear how resource intensive this can be, a typical lighthouse report takes 5-10 seconds to generate so it may be necessary to have a higher than usual `minimum_collection_interval`.
```

ruthnaebeck commented on lighthouse/README.md:
```diff
@@ -0,0 +1,81 @@
+
+### Requirements
+ 
+1. Check you have Node and npm installed
```
```suggestion
1. Check you have Node and NPM installed:
```

ruthnaebeck commented on lighthouse/README.md:
````diff
@@ -0,0 +1,81 @@
+```
+Lighthouse requires Node 8 LTS (8.9) or later.
+
+If not, [install Node and npm][5]
````
```suggestion
If not, [install Node and npm][5].
```

ruthnaebeck commented on lighthouse/README.md:
```diff
@@ -0,0 +1,81 @@
+
+If not, [install Node and npm][5]
+
+2. [Install Lighthouse][6]
```
```suggestion
2. [Install Lighthouse][6]:
```

ruthnaebeck commented on lighthouse/README.md:
````diff
@@ -0,0 +1,81 @@
+└── puppeteer@1.12.2
+```
+
+If not, install Chrome or [Puppeteer][7]
````
```suggestion
If not, install Chrome or [Puppeteer][7]:
```

ruthnaebeck commented on lighthouse/README.md:
```diff
@@ -0,0 +1,81 @@
+[6]: https://github.com/GoogleChrome/lighthouse
+[7]: https://github.com/GoogleChrome/puppeteer
+[8]: https://github.com/DataDog/integrations-extras/blob/master/lighthouse/datadog_checks/lighthouse/metadata.csv
+[9]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-status-and-information
```
```suggestion
[9]: https://github.com/DataDog/integrations-extras/blob/master/lighthouse/datadog_checks/lighthouse/metadata.csv
```

ruthnaebeck commented on lighthouse/README.md:
```diff
@@ -0,0 +1,81 @@
+[5]: https://nodejs.org/en/download/
+[6]: https://github.com/GoogleChrome/lighthouse
+[7]: https://github.com/GoogleChrome/puppeteer
+[8]: https://github.com/DataDog/integrations-extras/blob/master/lighthouse/datadog_checks/lighthouse/metadata.csv
```
```suggestion
[8]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-status-and-information
```

ericmustin commented:
@ruthnaebeck thanks for the additional feedback! added those suggestions, please let me know if there's anything i can do to help here

dabcoder commented on lighthouse/datadog_checks/lighthouse/data/conf.yaml.example:
```diff
@@ -0,0 +1,26 @@
+
+    # tags: 
+    #   - example_tag
+    #   - another_example_tag:an_example_value
```
Nit: new line

ericmustin replied:
👍

dabcoder commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,82 @@
+
+class LighthouseCheck(AgentCheck):
+    def check(self, instance):
+        EXPECTED_RESPONSE_CODE = "NO_ERROR"
```
Global variable should be defined outside of the class, you can put it above the class declaration.
Also, is that what Lightouse returns as response code? No HTTP code?

ericmustin replied:
Moved global variable outside the class declaration, ty.  oddly in regards to your other question, yes, lighthouse is just a command line program (it's making http requests under the hood) so it won't return the http response code from the underlying url it's checking, it just generates a JSON that either contains "no error" or details about the error in plain text. There's a variety of different possible errors but only one "success" message, which is "no error"

dabcoder commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,82 @@
+            tags = []
+
+        tags.append("lighthouse_url:%s" % lighthouse_url)
+        tags.append("lighthouse_name:%s" % lighthouse_name)
```
Nit too: you may want to use `.format` instead of `%s` here as you did in the tests.

ericmustin replied:
good call updated, ty

dabcoder commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,82 @@
+            return
+        # add tags
+        try:
+            tags = instance['tags']
```
If `tags` is `None` (users have not defined any custom tags) this will throw a `NoneType` Exception. You can use `instance.get('tags', [])` instead (empty list as default value). Example of this in he disk check: https://github.com/DataDog/integrations-core/blob/master/disk/datadog_checks/disk/disk.py#L56.
Then I don't think you'd need to check if `tags` is a list.

ericmustin replied:
updated to .get('tags',[]) but i think i ought to keep the check if it's a list in case the user has formatted the config incorrectly, just to be defensive?

dabcoder replied:
We don't do it in other checks as far as I can see, but I don't think it hurts to check that indeed.

dabcoder commented on lighthouse/metadata.csv:
```diff
@@ -0,0 +1,6 @@
+lighthouse.best_practices,gauge,,percent,,The google chrome lighthouse score for best_practices,0,lighthouse,best_practices
+lighthouse.performance,gauge,,percent,,The google chrome lighthouse score for performance,0,lighthouse,performance
+lighthouse.pwa,gauge,,percent,,The google chrome lighthouse score for pwa,0,lighthouse,pwa
+lighthouse.seo,gauge,,percent,,The google chrome lighthouse score for seo,0,lighthouse,seo
```
New line too.

dabcoder commented on lighthouse/tox.ini:
```diff
@@ -0,0 +1,29 @@
+skip_missing_interpreters = true
+basepython = py27
+envlist =
+    {py27,py36}-lighthouse
```
Should be `py37` there now.

ericmustin replied:
👍

FlorianVeaux commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+# the following try/except block will make the custom check compatible with any Agent version
+try:
```
You could directly use 
```python
from datadog_checks.base import AgentCheck
```
as there is already a similar compatibility layer.

ericmustin replied:
will update, ty

FlorianVeaux commented on lighthouse/datadog_checks/lighthouse/data/conf.yaml.example:
```diff
@@ -0,0 +1,26 @@
+init_config:
```
We try to follow a single format for all configuration file described [here](https://docs.datadoghq.com/developers/integrations/new_check_howto/#configuration). As you can see the testing pipeline now fails to validate this configuration file while it was working before. This is due to caching on our side, the tests were using an old version of the configuration validator which was more permissive.  Here is a suggested fix for it:

```
init_config:

instances:
    ## @param min_collection_interval - integer - required
    ## The lighthouse check can be resource intensive
    ## and can take ~10 seconds per instance, this has only been tested
    ## with a min collection interval of 60 seconds for a few instances
    #
  - min_collection_interval: 60

    ## @param url - string - required
    ## url of the website to get lighthouse metrics of
    ## ex: "https://www.datadoghq.com"
    #
    url: <URL_PLACEHOLDER>

    ## @param name - string - required
    ## name of your website.
    ## ex: "datadog_homepage"
    #
    name: <NAME_PLACEHOLDER>

    ## @param tags - list of key:value elements - optional
    ## List of tags to attach to every metric, event and service check emitted by this integration.
    ##
    ## Learn more about tagging: https://docs.datadoghq.com/tagging/
    #
    # tags:
    #   - <KEY_1>:<VALUE_1>
    #   - <KEY_2>:<VALUE_2>
```

ericmustin replied:
thank you for this, apologies on the formatting originally, cleaned up and updated

FlorianVeaux commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+import json
+
+# content of the special variable __version__ will be shown in the Agent status page
+__version__ = "1.0.0"
```
You can remove those lines, the `__version__` variable in your `__about__.py` file is the one being used for an integration.

FlorianVeaux commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+            if type(tags) != list:
+                self.log.warn('The tags list in the lighthouse check is not configured properly')
+                tags = []
+        except KeyError:
```
You can also remove this try catch block, as `instance.get` is fail safe

FlorianVeaux commented on lighthouse/datadog_checks/lighthouse/__about__.py:
```diff
@@ -0,0 +1 @@
+__version__ = '0.0.1'
```
You might want to bump to 1.0.0

AlexandreYang commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+            tags = []
+
+        tags.append("lighthouse_url:{0}".format(lighthouse_url))
+        tags.append("lighthouse_name:{0}".format(lighthouse_name))
```
Usually, there is no need to add prefix to tags.

- `lighthouse_name` could be `name`
- `lighthouse_url ` could be `url`

ericmustin replied:
makes sense, updated

AlexandreYang commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+            self.log.error("missing instance url or name")
+            raise CheckException("missing lighthouse instance url or name, please fix yaml")
+
+        cmd = ["/usr/local/bin/lighthouse",
```
```suggestion
        cmd = ["lighthouse",
```

ericmustin replied:
good catch, thanks updated.

ericmustin commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+        self.gauge("lighthouse.pwa", score_pwa, tags=tags)
+        self.gauge("lighthouse.seo", score_seo, tags=tags)
+
+    def _get_lighthouse_report(command, logger, raise_on_empty=False):
```
Thank you, apologies my python is pretty remedial, thank you for the example on this

ericmustin commented on lighthouse/datadog_checks/lighthouse/lighthouse.py:
```diff
@@ -0,0 +1,83 @@
+            score_best_practices = round(data["categories"]["best-practices"]["score"] * 100)
+            score_performance = round(data["categories"]["performance"]["score"] * 100)
+            score_pwa = round(data["categories"]["pwa"]["score"] * 100)
+            score_seo = round(data["categories"]["seo"]["score"] * 100)
```
This may be due to my level of python knowledge, but I was running into issues where the scores were being represented as, for example, 57.000000000000001, or something along those lines. So the rounding way my hacky way of fixing that. Fwiw these scores are only represented as integers, so in practice it shouldn't impact any of the data precision.

If there's a more pythonic way of handling this I would be happy to change it, open to suggestions and appreciate you taking a look.

AlexandreYang replied:
> This may be due to my level of python knowledge, but I was running into issues where the scores were being represented as, for example, 57.000000000000001, or something along those lines.

This is due to the way python (same for many other languages) represent floats, some decimal values cannot be represented exactly. Here are few links, if you want to dig into this :) https://docs.python.org/3/tutorial/floatingpoint.html, see also IEEE-754.

> Fwiw these scores are only represented as integers, so in practice it shouldn't impact any of the data precision.

Should be fine, in that case.

ericmustin commented on lighthouse/logos/README.md:
```diff
@@ -0,0 +1,19 @@
+# Logos
```
Updated `saas_logos-small.png` , oddly avatars-bot.png looks to be at the appropriate scale to me (128x128) ? Lmk if I have something wrong here

AlexandreYang replied:
I got this on github image preview, seems only the middle one is correctly scaled:

![image](https://user-images.githubusercontent.com/49917914/59587292-3c97c280-90e5-11e9-9c6c-a01d07739f0f.png)

ericmustin replied:
weird, i'll resize/reupload these, thanks for taking a look

ericmustin commented:
Updated PR to fit most recent integration conventions per suggestions, as well as added the feedback suggested by @AlexandreYang @FlorianVeaux @dabcoder @ruthnaebeck , and updated specs to reflect changes. Thank you all for the feedback, I had two minor clarifications/questions on feedback suggested by @AlexandreYang if you wouldn't mind taking a quick look.

Generally I appreciate everyone taking a look here and excited to get this out the door so lmk if anything else is needed from my end to push this over the finish line, cheers!

AlexandreYang commented on lighthouse/logos/README.md:
```diff
@@ -0,0 +1,19 @@
+#### `avatars-bot.png` (128 × 128)
+
+- Event stream at `/event/stream`
+- Notification icons at `/report/monitor`
```
Seems the issue is still there :) Here is the diff of the related Github commit.

![image](https://user-images.githubusercontent.com/49917914/59593005-6b1b9a80-90f1-11e9-9ff2-db742602454a.png)

ericmustin replied:
oy, ok I guess I am a bit confused? according to the readme doc and spot checking vs other integrations, these pixel width/height seem correct to me? ie
`saas_logos-small.png`: 120px Width / 60px Height
`saas_logos-bot.png`: 200px Width / 128px Height
`avatars-bot.png`: 128px Width / 128px Height

AlexandreYang replied:
Sorry, I wasn't clear on my previous comment. The size fine, but the images seems to be squeezed horizontally or vertically somehow, not 1:1 scaling.

ericmustin replied:
ahh, i see, apologies on that my eye for visual is poor. I found some assets that scaled a little more appropriately, pushed them up in last commit, the squeezed effect should be gone if you want to take a look. Thanks again appreciate the patience here.

AlexandreYang replied:
`lighthouse/logos/saas_logos-small.png` seems still a bit squeezed :)

![image](https://user-images.githubusercontent.com/49917914/59682060-b4ddb100-91d5-11e9-9f6e-49b633d112a4.png)

ericmustin replied:
oy, alrighty i updated again, i am almost positive this one is pixel perfect
